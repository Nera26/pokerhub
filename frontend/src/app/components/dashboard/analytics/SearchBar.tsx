'use client';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';
import type { AuditLogType } from '@shared/types';

interface Props {
  search: string;
  setSearch: (v: string) => void;
  type: 'all' | AuditLogType;
  setType: (v: 'all' | AuditLogType) => void;
  onSubmit: () => void;
}

export default function SearchBar({
  search,
  setSearch,
  type,
  setType,
  onSubmit,
}: Props) {
  return (
    <div className="bg-card-bg p-6 rounded-2xl shadow-[0_4px_8px_rgba(0,0,0,0.3)]">
      <h3 className="text-lg font-bold mb-4">Search Logs</h3>
      <div className="flex gap-3">
        <div className="flex-1">
          <input
            type="text"
            placeholder="Search by description, user, or event type..."
            value={search}
            onChange={(e) => setSearch(e.target.value)}
            onKeyDown={(e) => {
              if (e.key === 'Enter') onSubmit();
            }}
            className="w-full bg-primary-bg border border-dark rounded-xl px-4 py-3 text-sm focus:border-accent-yellow focus:outline-none"
          />
        </div>
        <select
          value={type}
          onChange={(e) => setType(e.target.value as 'all' | AuditLogType)}
          className="bg-primary-bg border border-dark rounded-xl px-4 py-3 text-sm"
        >
          <option value="all">All Types</option>
          <option value="Login">Login</option>
          <option value="Table Event">Table Event</option>
          <option value="Broadcast">Broadcast</option>
          <option value="Error">Error</option>
        </select>
        <button
          onClick={onSubmit}
          className="bg-accent-green px-6 py-3 rounded-xl font-semibold hover:shadow-[0_0_20px_rgba(28,139,76,0.3)]"
          title="Search"
        >
          <FontAwesomeIcon icon={faSearch} />
        </button>
      </div>
    </div>
  );
}
