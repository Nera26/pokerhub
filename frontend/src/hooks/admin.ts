'use client';

import type { ZodSchema } from 'zod';
import { createGetHook } from './useApiQuery';
import {
  SecurityAlertsResponseSchema,
  type AlertItem,
  AdminEventsResponseSchema,
  type AdminEvent,
} from '@shared/types';

export function createAdminGetHook<T>(path: string, schema: ZodSchema<T>) {
  return createGetHook<T>(path, schema);
}

export const useAuditAlerts = createAdminGetHook<AlertItem[]>(
  '/api/admin/security-alerts',
  SecurityAlertsResponseSchema,
);

export const useAdminEvents = createAdminGetHook<AdminEvent[]>(
  '/api/admin/events',
  AdminEventsResponseSchema,
);
