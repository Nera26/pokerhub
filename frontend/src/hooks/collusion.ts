'use client';

import {
  useQuery,
  useQueries,
  useMutation,
  useQueryClient,
} from '@tanstack/react-query';
import {
  listFlaggedSessions,
  getActionHistory,
  applyAction,
} from '@/lib/api/collusion';
import type {
  FlaggedSession,
  ReviewAction,
  ReviewActionLog,
} from '@shared/types';

function nextAction(status: FlaggedSession['status']): ReviewAction | null {
  switch (status) {
    case 'flagged':
      return 'warn';
    case 'warn':
      return 'restrict';
    case 'restrict':
      return 'ban';
    default:
      return null;
  }
}

export function useFlaggedSessions(token?: string) {
  return useQuery({
    queryKey: ['flagged-sessions', token],
    queryFn: () => listFlaggedSessions(token!),
    enabled: !!token,
  });
}

export function useActionHistory(sessions: FlaggedSession[], token?: string) {
  const results = useQueries({
    queries: sessions.map((s) => ({
      queryKey: ['collusion-history', token, s.id],
      queryFn: () => getActionHistory(s.id, token!),
      enabled: !!token,
    })),
  });
  return sessions.reduce<Record<string, ReviewActionLog[]>>(
    (acc, sess, idx) => {
      acc[sess.id] = results[idx]?.data ?? [];
      return acc;
    },
    {},
  );
}

export function useApplyCollusionAction(token?: string, reviewerId?: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({
      id,
      status,
    }: {
      id: string;
      status: FlaggedSession['status'];
    }) => {
      const action = nextAction(status);
      if (!token || !reviewerId || !action) {
        throw new Error('Invalid action');
      }
      return applyAction(id, action, token, reviewerId);
    },
    onMutate: async ({ id, status }) => {
      const action = nextAction(status);
      if (!action || !token || !reviewerId) return;
      const sessionsKey = ['flagged-sessions', token];
      const historyKey = ['collusion-history', token, id];

      await queryClient.cancelQueries({ queryKey: sessionsKey });
      await queryClient.cancelQueries({ queryKey: historyKey });

      const prevSessions =
        queryClient.getQueryData<FlaggedSession[]>(sessionsKey);
      const prevHistory =
        queryClient.getQueryData<ReviewActionLog[]>(historyKey);

      const optimistic: ReviewActionLog = {
        action,
        timestamp: Date.now(),
        reviewerId,
      };

      queryClient.setQueryData<FlaggedSession[]>(sessionsKey, (old = []) =>
        old.map((s) => (s.id === id ? { ...s, status: action } : s)),
      );

      queryClient.setQueryData<ReviewActionLog[]>(historyKey, (old = []) => [
        ...old,
        optimistic,
      ]);

      return { prevSessions, prevHistory };
    },
    onError: (_err, { id }, ctx) => {
      const sessionsKey = ['flagged-sessions', token];
      const historyKey = ['collusion-history', token, id];
      if (ctx?.prevSessions) {
        queryClient.setQueryData(sessionsKey, ctx.prevSessions);
      }
      if (ctx?.prevHistory) {
        queryClient.setQueryData(historyKey, ctx.prevHistory);
      }
    },
    onSuccess: (confirmed, { id }) => {
      const historyKey = ['collusion-history', token, id];
      queryClient.setQueryData<ReviewActionLog[]>(historyKey, (old = []) => [
        ...old.slice(0, -1),
        confirmed,
      ]);
    },
  });
}

export { nextAction };
