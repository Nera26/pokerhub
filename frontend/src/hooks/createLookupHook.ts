'use client';

import type { ZodSchema } from 'zod';
import { createGetHook } from './useApiQuery';

export function createLookupHook<T>(path: string, schema: ZodSchema<T>) {
  return createGetHook<T>(path, schema);
}

export default createLookupHook;
