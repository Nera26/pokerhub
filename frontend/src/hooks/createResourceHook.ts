'use client';

import { createQueryHook } from './useApiQuery';
import type { apiClient } from '@/lib/api/client';

export const createResourceHook = <T>(
  key: string,
  fetcher: (
    client: typeof apiClient,
    opts: { signal?: AbortSignal },
  ) => Promise<T>,
  label: string = key.replace(/-/g, ' '),
) => createQueryHook<T>(key, fetcher, label);
