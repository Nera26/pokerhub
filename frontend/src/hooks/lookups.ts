'use client';

import { createGetHook } from './useApiQuery';
import {
  BroadcastTypesResponseSchema,
  type BroadcastTypesResponse,
  AuditLogTypesResponseSchema,
  type AuditLogTypesResponse,
  LogTypeClassesSchema,
  type LogTypeClasses,
} from '@shared/types';

export const useBroadcastTypes = createGetHook<BroadcastTypesResponse>(
  '/api/broadcasts/types',
  BroadcastTypesResponseSchema,
);

export const useAuditLogTypes = createGetHook<AuditLogTypesResponse>(
  '/api/analytics/log-types',
  AuditLogTypesResponseSchema,
);

export const useAuditLogTypeClasses = createGetHook<LogTypeClasses>(
  '/api/analytics/log-types/classes',
  LogTypeClassesSchema,
);
