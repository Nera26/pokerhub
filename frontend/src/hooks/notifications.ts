'use client';

import { useQuery, type UseQueryOptions } from '@tanstack/react-query';
import {
  fetchNotifications,
  fetchUnreadCount,
  markAllNotificationsRead,
  markNotificationRead,
  fetchNotificationFilters,
} from '@/lib/api/notifications';
import type { ApiError, NotificationFilter } from '@/lib/api/notifications';
import type { NotificationsResponse } from '@shared/types';
import { useInvalidateMutationWithToast } from './useInvalidateMutationWithToast';

export function useNotifications(
  options?: Omit<
    UseQueryOptions<NotificationsResponse & { unread: number }, ApiError>,
    'queryKey' | 'queryFn'
  >,
) {
  return useQuery<NotificationsResponse & { unread: number }, ApiError>({
    queryKey: ['notifications'],
    queryFn: async ({ signal }) => {
      const [list, unread] = await Promise.all([
        fetchNotifications({ signal }),
        fetchUnreadCount({ signal }),
      ]);
      return { ...list, unread: unread.count };
    },
    ...options,
  });
}

export function useNotificationFilters(
  options?: Omit<
    UseQueryOptions<NotificationFilter[], ApiError>,
    'queryKey' | 'queryFn'
  >,
) {
  return useQuery<NotificationFilter[], ApiError>({
    queryKey: ['notificationFilters'],
    queryFn: ({ signal }) => fetchNotificationFilters({ signal }),
    ...options,
  });
}

export function useMarkAllRead() {
  return useInvalidateMutationWithToast({
    mutationFn: markAllNotificationsRead,
    queryKey: ['notifications'],
    update: (previous) => ({
      ...previous,
      notifications: previous.notifications.map((n) => ({
        ...n,
        read: true,
      })),
    }),
  });
}

export function useMarkRead() {
  return useInvalidateMutationWithToast({
    mutationFn: (id: string) => markNotificationRead(id),
    queryKey: ['notifications'],
    update: (previous, id) => ({
      ...previous,
      notifications: previous.notifications.map((n) =>
        n.id === id ? { ...n, read: true } : n,
      ),
    }),
  });
}
