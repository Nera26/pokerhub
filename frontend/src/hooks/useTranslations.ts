'use client';

import { createQueryHook } from './createQueryHook';
import { fetchTranslations } from '@/lib/api/translations';

export const useTranslations = createQueryHook<Record<string, string>, string>(
  'translations',
  (_client, lang, opts) =>
    fetchTranslations(lang, { signal: opts.signal }).then((r) => r.messages),
  'translations',
);
