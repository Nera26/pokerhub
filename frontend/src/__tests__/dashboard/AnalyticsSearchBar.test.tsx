import { render, screen } from '@testing-library/react';
import { useLocale } from 'next-intl';
import { useTranslations } from '@/hooks/useTranslations';
import { useAuditLogTypes } from '@/hooks/lookups';
import SearchBar from '@/app/components/dashboard/analytics/SearchBar';

jest.mock('next-intl', () => ({
  useLocale: jest.fn(),
}));

jest.mock('@/hooks/useTranslations', () => ({
  useTranslations: jest.fn(),
}));

jest.mock('@/hooks/lookups', () => ({
  useAuditLogTypes: jest.fn(),
}));

const mockedUseLocale = jest.mocked(useLocale);
const mockedUseTranslations = jest.mocked(useTranslations);
const mockedUseAuditLogTypes = jest.mocked(useAuditLogTypes);

describe('Analytics SearchBar translations', () => {
  const renderComponent = () =>
    render(
      <SearchBar
        search=""
        setSearch={jest.fn()}
        type="all"
        setType={jest.fn()}
        onSubmit={jest.fn()}
      />,
    );

  beforeEach(() => {
    jest.resetAllMocks();
    mockedUseLocale.mockReturnValue('en');
    mockedUseAuditLogTypes.mockReturnValue({
      data: { types: ['security'] },
      isLoading: false,
      isError: false,
    } as unknown as ReturnType<typeof useAuditLogTypes>);
    mockedUseTranslations.mockReturnValue({
      data: undefined,
    } as unknown as ReturnType<typeof useTranslations>);
  });

  it('uses translated placeholder text when available', () => {
    mockedUseTranslations.mockReturnValue({
      data: { auditSearchPlaceholder: 'Custom audit search' },
    } as unknown as ReturnType<typeof useTranslations>);

    renderComponent();

    expect(screen.getByRole('textbox')).toHaveAttribute(
      'placeholder',
      'Custom audit search',
    );
  });

  it('falls back to default placeholder when translation is missing', () => {
    const { container } = renderComponent();

    expect(screen.getByRole('textbox')).toHaveAttribute(
      'placeholder',
      'Searchâ€¦',
    );
    expect(container).toMatchSnapshot();
  });
});
