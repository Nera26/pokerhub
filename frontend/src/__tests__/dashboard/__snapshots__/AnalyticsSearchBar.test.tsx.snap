// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics SearchBar translations falls back to default placeholder when translation is missing 1`] = `
<div>
  <div
    class="bg-card-bg p-6 rounded-2xl shadow-[0_4px_8px_rgba(0,0,0,0.3)]"
  >
    <h3
      class="text-lg font-bold mb-4"
    >
      Search Logs
    </h3>
    <div
      class="flex gap-3"
    >
      <div
        class="flex-1"
      >
        <div
          class="relative"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-magnifying-glass absolute left-3 top-1/2 -translate-y-1/2 text-text-secondary"
            data-icon="magnifying-glass"
            data-prefix="fas"
            role="img"
            viewBox="0 0 512 512"
          >
            <path
              d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376C296.3 401.1 253.9 416 208 416 93.1 416 0 322.9 0 208S93.1 0 208 0 416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
              fill="currentColor"
            />
          </svg>
          <input
            class="w-full bg-primary-bg border border-dark rounded-xl pl-10 pr-4 text-sm text-text-primary focus:border-accent-yellow focus:outline-none py-3"
            placeholder="Searchâ€¦"
            value=""
          />
        </div>
      </div>
      <select
        class="bg-primary-bg border border-dark rounded-xl px-4 py-3 text-sm"
      >
        <option
          value="all"
        >
          All Types
        </option>
        <option
          value="security"
        >
          security
        </option>
      </select>
      <button
        class="bg-accent-green px-6 py-3 rounded-xl font-semibold hover:shadow-[0_0_20px_rgba(28,139,76,0.3)]"
        title="Search"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass "
          data-icon="magnifying-glass"
          data-prefix="fas"
          role="img"
          viewBox="0 0 512 512"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376C296.3 401.1 253.9 416 208 416 93.1 416 0 322.9 0 208S93.1 0 208 0 416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
